File: app/src/main/java/com/yibao/music/fragment/dialogfrag/CrashSheetDialog.java
package com.yibao.music.fragment.dialogfrag;

import android.content.Context;
import android.content.Intent;
import android.content.pm.PackageInfo;
import android.content.pm.PackageManager;
import android.net.Uri;
import android.os.Build;
import android.os.Bundle;
import android.support.design.widget.BottomSheetDialog;
import android.support.v4.content.FileProvider;
import android.support.v7.widget.LinearLayoutManager;
import android.support.v7.widget.RecyclerView;
import android.view.LayoutInflater;
import android.view.View;
import android.view.Window;
import android.view.WindowManager;
import android.widget.LinearLayout;
import android.widget.TextView;

import com.yibao.music.R;
import com.yibao.music.adapter.CrashAdapter;
import com.yibao.music.base.factory.RecyclerFactory;
import com.yibao.music.util.Constants;
import com.yibao.music.util.FileUtil;

import java.io.File;
import java.util.List;

/**
 * Des：${TODO}
 * Time:2017/8/22 14:11
 *
 * @author Stran
 */
public class CrashSheetDialog {
    private Context mContext;
    private String mPackageName;
    private File mFiles;

    public static CrashSheetDialog newInstance() {
        return new CrashSheetDialog();
    }

    public void getBottomDialog(Context context) {
        this.mContext = context;
        mPackageName = mContext.getPackageName();
        BottomSheetDialog dialog = new BottomSheetDialog(context);
        View view = LayoutInflater.from(context)
                .inflate(R.layout.crash_bottom_sheet_dialog, null);
        init(dialog, view);
        dialog.show();
    }


    private void init(BottomSheetDialog dialog, View view) {
        dialog.setContentView(view);
        dialog.setCancelable(true);
        Window window = dialog.getWindow();
        if (window != null) {
            window.addFlags(WindowManager.LayoutParams.FLAG_TRANSLUCENT_STATUS);
        }
        dialog.setCanceledOnTouchOutside(true);
        LinearLayout rootView = view.findViewById(R.id.root_crash);
        // 读取CrashLog
        mFiles = new File(Constants.CRASH_LOG_PATH);
        if (mFiles.exists()) {
            File[] array = mFiles.listFiles();
            CrashAdapter crashAdapter = new CrashAdapter(array);
            RecyclerView recyclerView = RecyclerFactory.creatRecyclerView(Constants.NUMBER_ONE, crashAdapter);
            crashAdapter.setItemClickListener(this::openCrashLog);
            rootView.addView(recyclerView);
            TextView tvTitle = view.findViewById(R.id.tv_crash_title);
            tvTitle.setOnClickListener(v -> backTop(recyclerView));
            tvTitle.setOnLongClickListener(v -> {
                FileUtil.deleteFile(mFiles);
                dialog.dismiss();
                return true;
            });
        }

    }

    private void openCrashLog(File crashFile) {
        Intent intent = new Intent();
        intent.addFlags(Intent.FLAG_ACTIVITY_NEW_TASK);
        intent.setAction(Intent.ACTION_VIEW);
        intent.setFlags(Intent.FLAG_GRANT_READ_URI_PERMISSION);
        Uri contentUri;
        if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.N) {
            contentUri = FileProvider.getUriForFile(mContext, mPackageName, crashFile);
        } else {
            contentUri = Uri.fromFile(crashFile);
        }
        // 默认用WPS打开日志
        if (isAvilible()) {
            Bundle bundle = new Bundle();
            bundle.putString(WpsModel.OPEN_MODE, WpsModel.READ_MODE);
            bundle.putBoolean(WpsModel.SEND_CLOSE_BROAD, false);
            bundle.putString(WpsModel.THIRD_PACKAGE, mPackageName);
            bundle.putBoolean(WpsModel.CLEAR_TRACE, true);
            // bundle.putBoolean(CLEAR_FILE, true); //关闭后删除打开文件
            intent.setClassName(WpsModel.NORMAL_PACKAGE, WpsModel.NORMAL);
            intent.setDataAndType(contentUri, "application/vnd.android.package-archive");
            intent.putExtras(bundle);
        } else {
            intent.setDataAndType(contentUri, Constants.DATA_TYPE_TXT);
        }
        mContext.startActivity(intent);


    }

    private boolean isAvilible() {
        String wpsPackageName = "cn.wps.moffice_eng";
        final PackageManager packageManager = mContext.getPackageManager();
        // 获取所有已安装程序的包信息
        List<PackageInfo> pinfo = packageManager.getInstalledPackages(0);
        for (int i = 0; i < pinfo.size(); i++) {
            if (pinfo.get(i).packageName.equalsIgnoreCase(wpsPackageName)) {
                return true;
            }
        }
        return false;
    }

    private void backTop(RecyclerView recyclerView) {
        CrashAdapter adapter = (CrashAdapter) recyclerView.getAdapter();
        LinearLayoutManager manager = (LinearLayoutManager) recyclerView.getLayoutManager();
        if (adapter != null && manager != null) {
            int positionForSection = adapter.getPositionForSection(0);
            manager.scrollToPositionWithOffset(positionForSection, 0);
        }
    }

    private class WpsModel {
        static final String OPEN_MODE = "OpenMode";
        static final String SEND_CLOSE_BROAD = "SendCloseBroad";
        static final String THIRD_PACKAGE = "ThirdPackage";
        static final String CLEAR_TRACE = "ClearTrace";
        static final String READ_MODE = "ReadMode";
        static final String NORMAL = "cn.wps.moffice.documentmanager.PreStartActivity2";
        static final String NORMAL_PACKAGE = "cn.wps.moffice_eng";

    }
}

File: app/src/main/java/com/yibao/music/fragment/dialogfrag/CrashSheetDialog.kt
package com.yibao.music.fragment.dialogfrag

import android.content.Context
import android.content.Intent
import android.net.Uri
import android.os.Build
import android.os.Bundle
import android.view.LayoutInflater
import android.view.WindowManager
import androidx.core.content.ContextCompat
import androidx.core.content.FileProvider
import androidx.recyclerview.widget.DividerItemDecoration
import androidx.recyclerview.widget.LinearLayoutManager
import androidx.recyclerview.widget.RecyclerView
import com.google.android.material.bottomsheet.BottomSheetDialog
import com.yibao.music.MusicApplication
import com.yibao.music.R
import com.yibao.music.adapter.CrashAdapter
import com.yibao.music.databinding.CrashBottomSheetDialogBinding
import com.yibao.music.util.Constant
import com.yibao.music.util.FileUtil
import java.io.File

/**
 * Des：${TODO}
 * Time:2017/8/22 14:11
 *
 * @author Stran
 */
class CrashSheetDialog {
    private var mContext: Context? = null
    private lateinit var mBinding: CrashBottomSheetDialogBinding
    private lateinit var mDialog: BottomSheetDialog

    fun getBottomDialog(context: Context?) {
        mContext = context
        mBinding = CrashBottomSheetDialogBinding.inflate(LayoutInflater.from(context), null, false)
        mDialog = BottomSheetDialog(context!!)
        mDialog.setContentView(mBinding.rootCrash)
        mDialog.setCancelable(true)
        mDialog.setCanceledOnTouchOutside(true)
        init()
        mDialog.show()
    }

    private fun init() {
        initRecyclerView(mBinding.recyclerCrashLog)
        // 读取CrashLog
        val crashDir = FileUtil.getCrashDir()
        if (crashDir.exists()) {
            val array = crashDir.listFiles().apply {
                this?.reverse()
            }
            val crashAdapter = CrashAdapter(array)
            mBinding.recyclerCrashLog.adapter = crashAdapter
            crashAdapter.setItemClickListener { crashFile: File -> openCrashLog(crashFile) }
            mBinding.tvCrashTitle.setOnClickListener { backTop(mBinding.recyclerCrashLog) }
            mBinding.tvCrashTitle.setOnLongClickListener {
                FileUtil.deleteFileDirectory(crashDir)
                mDialog.dismiss()
                true
            }
        }

    }


    private fun openCrashLog(crashFile: File) {
        val packageName = mContext!!.packageName
        val intent = Intent()
        intent.addFlags(Intent.FLAG_ACTIVITY_NEW_TASK)
        intent.action = Intent.ACTION_VIEW
        intent.flags = Intent.FLAG_GRANT_READ_URI_PERMISSION
        val contentUri = if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.N) {
            FileProvider.getUriForFile(mContext!!, packageName, crashFile)
        } else {
            Uri.fromFile(crashFile)
        }
        // 默认用WPS打开日志
        if (isInstallWps()) {
            val bundle = Bundle()
            bundle.putString(WpsModel.OPEN_MODE, WpsModel.READ_MODE)
            bundle.putBoolean(WpsModel.SEND_CLOSE_BROAD, false)
            bundle.putString(WpsModel.THIRD_PACKAGE, packageName)
            bundle.putBoolean(WpsModel.CLEAR_TRACE, true)
            // bundle.putBoolean(CLEAR_FILE, true); //关闭后删除打开文件
            intent.setClassName(WpsModel.NORMAL_PACKAGE, WpsModel.NORMAL)
            intent.setDataAndType(contentUri, "application/vnd.android.package-archive")
            intent.putExtras(bundle)
        } else {
            intent.setDataAndType(contentUri, Constant.DATA_TYPE_TXT)
        }
        mContext!!.startActivity(intent)
    }

    // 获取所有已安装程序的包信息
    private fun isInstallWps(): Boolean {
        val wpsPackageName = "cn.wps.moffice_eng"
        val packageManager = mContext!!.packageManager
        // 获取所有已安装程序的包信息
        val info = packageManager.getInstalledPackages(0)
        for (i in info.indices) {
            if (info[i].packageName == wpsPackageName) {
                return true
            }
        }
        return false
    }

    private fun backTop(recyclerView: RecyclerView) {
        val adapter = recyclerView.adapter as CrashAdapter?
        val manager = recyclerView.layoutManager as LinearLayoutManager?
        if (adapter != null && manager != null) {
            val positionForSection = adapter.getPositionForSection(0)
            manager.scrollToPositionWithOffset(positionForSection, 0)
        }
    }

    private fun initRecyclerView(recyclerView: RecyclerView) {
        val manager = LinearLayoutManager(MusicApplication.getInstance())
        manager.orientation = LinearLayoutManager.VERTICAL
        recyclerView.isVerticalScrollBarEnabled = true
        recyclerView.layoutManager = manager
        val divider = DividerItemDecoration(mContext, DividerItemDecoration.VERTICAL)
        ContextCompat.getDrawable(mContext!!, R.drawable.shape_item_decoration)
            ?.let { divider.setDrawable(it) }
        recyclerView.addItemDecoration(divider)
    }

    private object WpsModel {
        const val OPEN_MODE = "OpenMode"
        const val SEND_CLOSE_BROAD = "SendCloseBroad"
        const val THIRD_PACKAGE = "ThirdPackage"
        const val CLEAR_TRACE = "ClearTrace"
        const val READ_MODE = "ReadMode"
        const val NORMAL = "cn.wps.moffice.documentmanager.PreStartActivity2"
        const val NORMAL_PACKAGE = "cn.wps.moffice_eng"
    }

    companion object {
        fun newInstance(): CrashSheetDialog {
            return CrashSheetDialog()
        }
    }
}